core/src/cgcloud/core/box.py

```diff
@@ -373,15 +378,120 @@ def prepare( self, ec2_keypair_globs, instance_type=None, image_ref=None,
+        spot_support = ec2_instance_types[instance_type].spot_availability
+        assert spot_support is True
+        # this means we are using a spot bid. We need a couple additional kwargs
+        self._optimize_bid(instance_type, options, kwargs)
+        kwargs['price'] = options['price']
+        kwargs['launch_group'] = self.ctx.namespace.replace("/", "") + "_launch_group"
+        try:  # There is no min/max count parameter for request_spot_instances, just count
+            kwargs['count'] = kwargs['max_count']
+            del kwargs['max_count']
+            del kwargs['min_count']
+        except KeyError:
+            pass
+
+    @staticmethod
+    def _choose_zone(zones, bid, spot_data):
+        """
+        returns the zone to put the spot request based on, in order of priority:
+           1) zones with prices currently under the bid
+           2) zones with the most stable price
+        :return: String representing the name of the selected zone
+
+        >>> from collections import namedtuple
+        >>> FauxHistory = namedtuple("FauxHistory",["price","availability_zone"])
+        >>> ZoneTuple = namedtuple("ZoneTuple", ["name"])
+        >>> zones =[ZoneTuple("us-west-2a"),ZoneTuple("us-west-2b")]
+        >>> spot_data=[FauxHistory(0.1,"us-west-2a"),FauxHistory(0.2,"us-west-2a"),FauxHistory(0.3,"us-west-2b"),FauxHistory(0.6,"us-west-2b")]
+        >>> Box._choose_zone(zones, 0.15, spot_data)
+        'us-west-2a'
+        >>> spot_data=[FauxHistory(0.3,"us-west-2a"),FauxHistory(0.2,"us-west-2a"),FauxHistory(0.1,"us-west-2b"),FauxHistory(0.6,"us-west-2b")]
+        >>> Box._choose_zone(zones, 0.15, spot_data)
+        'us-west-2b'
+        >>> spot_data=[FauxHistory(0.1,"us-west-2a"),FauxHistory(0.7,"us-west-2a"),FauxHistory(0.1,"us-west-2b"),FauxHistory(0.6,"us-west-2b")]
+        >>> Box._choose_zone(zones, 0.15, spot_data)
+        'us-west-2b'
+       """
+
+        ZoneTuple = namedtuple("ZoneTuple", ["name","price_deviation"])
+        def _sort_zones(bid, spot_data, zones):
+            # returns two lists of tuples of form: [ (zone.name, std_deviation)...]
+            # one for zones over the bid price and one for zones under bid price. Each are sorted by
+            # increasing standard deviation values
+            markets_under_bid, markets_over_bid = [], []
+            for zone in zones:
+                zlist = filter(lambda spot_history: spot_history.availability_zone == zone.name, spot_data)
+                std_deviation = std_dev([spot_history.price for spot_history in zlist])
+                # http://stackoverflow.com/a/12135169
+                (markets_over_bid, markets_under_bid)[zlist[0].price < bid].append(ZoneTuple(zone.name,std_deviation))
+            # sort each list in place by increasing standard deviation
+            map(lambda list: list.sort(key=attrgetter("price_deviation")), [markets_over_bid, markets_under_bid])
```

Comment:
Use min() and max() to get an extreme. Not sort()[0]. The former is O(n), the latter O(n log n) at best.